brave/src/main/java/brave/Tag.java

```diff
@@ -0,0 +1,154 @@
+   * @since 5.11
+   */
+  @Nullable protected abstract String parseValue(I input, @Nullable TraceContext context);
+
+  /**
+   * {@linkplain ScopedSpan#tag(String, String) Tags} the value parsed from the {@code input} when
+   * non-null and the {@linkplain ScopedSpan#isNoop() span is not no-op}.
+   *
+   * @since 5.11
+   */
+  public final void tag(I input, ScopedSpan span) {
+    if (input == null) throw new NullPointerException("input == null");
+    if (span == null) throw new NullPointerException("span == null");
+    tag(span, input, span.context());
+  }
+
+  /**
+   * {@linkplain SpanCustomizer#tag(String, String) Tags} the value parsed from the {@code input}
+   * when non-null and the span is not no-op.
+   *
+   * @since 5.11
+   */
+  // ex void parse(HttpRequest request, TraceContext context, SpanCustomizer span);
+  public final void tag(I input, TraceContext context, SpanCustomizer span) {
+    if (input == null) throw new NullPointerException("input == null");
+    if (context == null) throw new NullPointerException("context == null");
+    if (span == null) throw new NullPointerException("span == null");
+    if (span == NoopSpanCustomizer.INSTANCE) return;
+    tag(span, input, context);
+  }
+
+  /**
+   * {@linkplain SpanCustomizer#tag(String, String) Tags} the value parsed from the {@code input}
+   * when non-null and the span is not no-op.
+   *
+   * @since 5.11
+   */
+  public final void tag(I input, SpanCustomizer span) {
+    if (input == null) throw new NullPointerException("input == null");
+    if (span == null) throw new NullPointerException("span == null");
+    TraceContext context = null;
+    if (span instanceof Span) {
+      Span asSpan = (Span) span;
+      if (asSpan.isNoop()) return;
+      context = asSpan.context();
+    } else if (span == NoopSpanCustomizer.INSTANCE) {
+      return;
+    }
+    tag(span, input, context);
+  }
+
+  /**
+   * {@linkplain MutableSpan#tag(String, String) Tags} the value parsed from the {@code input} when
+   * non-null.
+   *
+   * @see FinishedSpanHandler#handle(TraceContext, MutableSpan)
+   * @since 5.11
+   */
+  public final void tag(I input, TraceContext context, MutableSpan span) {
+    if (input == null) throw new NullPointerException("input == null");
+    if (span == null) throw new NullPointerException("span == null");
+    tag(span, input, context);
+  }
+
+  final String key;
+
+  /** @since 5.11 */
+  protected Tag(String key) {
+    this.key = validateNonEmpty("key", key);
+  }
+
+  @Override public String toString() {
+    return getClass().getSimpleName() + "{" + key + "}";
+  }
+
+  final void tag(Object span, I input, @Nullable TraceContext context) {
+    String value;
+    try {
+      value = parseValue(input, context);
+    } catch (Throwable e) {
+      propagateIfFatal(e);
+      Platform.get().log("Error parsing tag value of input %s", input, e);
+      return;
+    }
+    if (value == null) return;
+    if (span instanceof SpanCustomizer) {
```

Comment:
I'm sure it must have been thought no need for detail - any thoughts on a shared interface?

Reply:
sure. here's the context

Span is indeed instanceof SpanCustomizer
ScopedSpan doesn't allow name overrides
MutableSpan doesn't have an implicit timestamp

Reply:
we could allow ScopedSpan to override its name.. now wouldn't be a bad time to do that..

Reply:
https://github.com/openzipkin/brave/pull/1134